contracts/SpokePoolV3Periphery.sol

```diff
@@ -14,78 +14,132 @@ import { IERC20Auth } from "./external/interfaces/IERC20Auth.sol";
+    bool private expectingPermit2Callback;
+
+    // Events
+    event SwapExecuted(
+        address exchange,
+        address indexed inputToken,
+        address indexed outputToken,
+        uint256 outputAmount,
+        SpokePoolV3PeripheryInterface.TransferType transferType
+    );
+
+    // Errors
+    error SwapFailed();
+    error UnsupportedTransferType();
 
     /**
-     * @notice Initialize the SpokePoolPeripheryProxy contract.
-     * @param _spokePoolPeriphery Address of the SpokePoolPeriphery contract that this proxy will call.
+     * @notice Constructs a new SwapProxy.
+     * @param _permit2 Address of the canonical permit2 contract.
      */
-    function initialize(SpokePoolV3Periphery _spokePoolPeriphery) external nonReentrant {
-        if (initialized) revert ContractInitialized();
-        initialized = true;
-        if (!address(_spokePoolPeriphery).isContract()) revert InvalidPeriphery();
-        spokePoolPeriphery = _spokePoolPeriphery;
+    constructor(address _permit2) {
+        require(_permit2 != address(0), "Permit2 cannot be zero address");
+        permit2 = IPermit2(_permit2);
     }
 
     /**
-     * @inheritdoc SpokePoolV3PeripheryProxyInterface
+     * @notice Executes a swap on the given exchange with the provided calldata.
+     * @param inputToken The token to swap from
+     * @param outputToken The token to swap to
+     * @param inputAmount The amount of input tokens to swap
+     * @param exchange The exchange to perform the swap
+     * @param transferType The method of transferring tokens to the exchange
+     * @param routerCalldata The calldata to execute on the exchange
+     * @return outputAmount The actual amount of output tokens received from the swap
      */
-    function swapAndBridge(SpokePoolV3PeripheryInterface.SwapAndDepositData calldata swapAndDepositData)
-        external
-        override
-        nonReentrant
-    {
-        _callSwapAndBridge(swapAndDepositData);
+    function performSwap(
+        address inputToken,
+        address outputToken,
+        uint256 inputAmount,
+        address exchange,
+        SpokePoolV3PeripheryInterface.TransferType transferType,
+        bytes calldata routerCalldata
+    ) external nonReentrant returns (uint256 outputAmount) {
+        // We'll return the final balance of output tokens
+
+        // The exchange will either receive funds from this contract via:
+        // 1. A direct approval to spend funds on this contract (TransferType.Approval),
+        // 2. A direct transfer of funds to the exchange (TransferType.Transfer), or
+        // 3. A permit2 approval (TransferType.Permit2Approval)
+        if (transferType == SpokePoolV3PeripheryInterface.TransferType.Approval) {
+            IERC20(inputToken).forceApprove(exchange, inputAmount);
+        } else if (transferType == SpokePoolV3PeripheryInterface.TransferType.Transfer) {
+            IERC20(inputToken).transfer(exchange, inputAmount);
```

Comment:
nit: use `safeTransfer` instead (I see we didn't do this the first time, either)?

Reply:
https://github.com/across-protocol/contracts/pull/1002/commits/e208d327ab195738b781153bf75dd1f5bdbf4960